pip/pip-269.md

```diff
@@ -0,0 +1,239 @@
+- If there has an in-flight reading.
+- If there is more than one **Occasional Read Position Change** action running.
+- Both above.
+
+---
+
+In the Multi Consumer Dispatcher, there has a mechanism that tries to fix the issue caused by the action `cursor.rewind` and **Sequentially Read** executed concurrently(referred to as Rewind After In-flight Reading). It works like this: if there has an in-flight reading when calling `cursor.rewind`, just do rewind after the in-flight reading is finished.` but this mechanism has the following drawbacks:
+- Makes a new issue lead to a new issue<sup>[5]</sup>.
+- Only applicable to `cursor.rewind`.
+- Only applicable to Multi Consumer Dispatcher.
+- If there is more than one in-flight reading(one Sequentially Read and some Message Replay Read might exist simultaneously), this mechanism can not work.
+
+# Goals
+- For the **Scenario-1** and **Scenario-2**(see Motivation), discard the in-flight reading(referred to as **Goal-1**).
+- For the **Scenario-3**(see Motivation), we should do these two things:
+    - Provide a way to guarantee the atomicity modification of both attributes of Cursor and attributes of Dispatcher(referred to as **Goal-2**).
+    - Prevent more than one **Occasional Read Position Change** executed at the same time(Current Proposal does not consider this optimization, there will be some bug fixes after this PIP)
+- Revert https://github.com/apache/pulsar/issues/6403
+- Remove the mechanism Rewind After In-flight Reading<sup>[5]</sup>
+
+# High-Level Design
+
+**For the Goal-1**: Cursor Epoch
+- Definition of a counter to record how many times the `read position` of the cursor has been changed by the **Occasional Read Position Change**, we call it Cursor Epoch. **Occasional Read Position Change** executed once, the Cursor Epoch increase once.
+- Every **Sequentially Read** or **Message Replay Read** carries the current Cursor Epoch.
+- Discard the reading and prevent the `read position` change if the epoch carried by the reading is smaller than the current Cursor Epoch when the reading complete
+
+**For the Goal-2**: Two-phase of Cursor Epoch increment
+- Mark the Cursor Epoch increase task is started.
+    - Set a new `read position`.
+    - Mark the cursor is modifying, at this moment, all the **Sequentially Read** and **Message Replay Read** will fail due to `CursorModifyingException`.
+- We can change attributes of the Cursor or attributes of the Dispatcher now.
+- End the Cursor Epoch increase task
+    - increase the epoch.
+    - remove the marker `modifying`.
+
+# Detailed Design
+
+**ManagedLedgerException.java**
+```java
+public static class CursorModifyingException extends ManagedLedgerException {};
+
+public static class CursorEpochConflictException extends ManagedLedgerException {};
+```
+
+<strong>ManagedCursorImpl.java</strong>(maybe instead `synchronized` to a `lock`)
```

Comment:
Please, let's not add more code to `ManagedCursorImpl` - it becomes a dumpster.
Let's try to concentrate the same domain into one class. 
First, I need to understand the high level design of goal 2 to be able to give good suggestions here.

Reply:
Good suggestion. already added a new component `CursorEpoch.java`